src/trusted/service_runtime/nacl_signal_common.c

```diff
@@ -134,7 +138,23 @@ void NaClSignalHandleUntrusted(int signal,
     SNPRINTF(tmp, sizeof(tmp), "\n** Signal %d from untrusted code: "
              "pc=%" NACL_PRIxNACL_REG "\n", signal, regs->prog_ctr);
     NaClSignalErrorMessage(tmp);
-    NaClExit((-signal) & 0xFF);
+
+    /*
+      * Unset the TLS variable so that if a crash occurs during thread
+      * teardown, the signal handler does not dereference a dangling
+      * NaClAppThread pointer.
+    */
+    NaClTlsSetCurrentThread(NULL);
+    // signal for teardown
+    if (natp->is_cage_mainthread) {
+      if (!natp->tearing_down) AddToFatalThreadTeardown(natp);
+      NaClUntrustedThreadSuspend(natp, 0);   
+    } else {
+      if (!natp->cage_mainthread->tearing_down) AddToFatalThreadTeardown(natp->cage_mainthread);
+    }
+    // wait here while we get cleaned up
+    while (1);
```

Comment:
At this point you should probably do a `pause()` instead of busy loop?

Reply:
That is the purpose of pause after all

Reply:
I'm almost certain I can't use pause() in a signal handler. My capabilities here are pretty minimal.

Reply:
Oh right makes sense